# Description

In this problem, we want to implement a perception learning algorithm. Please write a simple python code to achieve the algorithm.

In [1]:
# import the module
import numpy as np

In [49]:
def train_weights(train: np.array, l_rate: float, n_epoch: int, weights: iter):
    for n in range(n_epoch):
        for d in range(len(train)):
            z = weights[0] + train[d][0] * weights[1] + train[d][1] * weights[2] + train[d][2] * weights[3]
            y = -1 if z < 0 else 1
            for w in range(4):
                weights[w] = weights[w] + l_rate * (train[d][3] - y) * (1 if w == 0 else train[d][w-1])
        print("epoch = {}, weights = {:.2f} {:.2f} {:.2f} {:.2f}".format(n+1, weights[0], weights[1], weights[2], weights[3]))
    

In [50]:
# This is the dataset we used. It is the same as shown in the 'perceptron.pdf'. 
# The first three numbers of each row are features, and the last number is the class label.
dataset = np.array([
    [1, 0, 0, -1],
    [1, 0, 1, 1],
    [1, 1, 0, 1],
    [1, 1, 1, 1],
    [0, 0, 1, -1],
    [0, 1, 0, -1],
    [0, 1, 1, 1],
    [0, 0, 0, -1]])

# Parameters
l_rate = 0.1 # learning rate
n_epoch = 10 # number of the epochs
init_weights = [0.5, 0.5, 0.5, 0.5] # TODO: implement the initial weights

# Call the function and print the results
train_weights(dataset, l_rate, n_epoch, init_weights)

epoch = 1, weights = -0.10 0.30 0.30 0.30
epoch = 2, weights = -0.30 0.10 0.30 0.30
epoch = 3, weights = -0.30 0.10 0.30 0.30
epoch = 4, weights = -0.30 0.10 0.30 0.30
epoch = 5, weights = -0.30 0.10 0.30 0.30
epoch = 6, weights = -0.30 0.10 0.30 0.30
epoch = 7, weights = -0.30 0.10 0.30 0.30
epoch = 8, weights = -0.30 0.10 0.30 0.30
epoch = 9, weights = -0.30 0.10 0.30 0.30
epoch = 10, weights = -0.30 0.10 0.30 0.30


# The outputs

For epoch = 10, learning rate = 0.1, initial weight = 0.5, the output should be displayed as following

epoch = 1, weights = -0.10 0.30 0.30 0.30 <br>
epoch = 2, weights = -0.30 0.10 0.30 0.30 <br>
epoch = 3, weights = -0.30 0.10 0.30 0.30 <br>
epoch = 4, weights = -0.30 0.10 0.30 0.30 <br>
epoch = 5, weights = -0.30 0.10 0.30 0.30 <br>
epoch = 6, weights = -0.30 0.10 0.30 0.30 <br>
epoch = 7, weights = -0.30 0.10 0.30 0.30 <br>
epoch = 8, weights = -0.30 0.10 0.30 0.30 <br>
epoch = 9, weights = -0.30 0.10 0.30 0.30 <br>
epoch = 10, weights = -0.30 0.10 0.30 0.30 <br>